New everything server

cliffhall opened the pull request:
## Description
* Refactor/extracted all features from `everything.ts` into a more modular and maintainable structure.

```
src/everything
     ├── index.ts
     ├── AGENTS.md
     ├── Dockerfile
     ├── package.json
     ├── tsconfig.json
     ├── docs
     │   ├── architecture.md
     │   ├── extension.md
     │   ├── features.md
     │   ├── how-it-works.md
     │   ├── instructions.md
     │   ├── startup.md
     │   └── structure.md
     ├── prompts
     │   ├── index.ts
     │   ├── args.ts
     │   ├── completions.ts
     │   ├── simple.ts
     │   └── resource.ts
     ├── resources
     │   ├── index.ts
     │   ├── files.ts
     │   ├── session.ts
     │   ├── subscriptions.ts
     │   └── templates.ts
     ├── server
     │   ├── index.ts
     │   ├── logging.ts
     │   └── roots.ts
     ├── tools
     │   ├── index.ts
     │   ├── echo.ts
     │   ├── get-annotated-message.ts
     │   ├── get-env.ts
     │   ├── get-resource-links.ts
     │   ├── get-resource-reference.ts
     │   ├── get-roots-list.ts
     │   ├── get-structured-content.ts
     │   ├── get-sum.ts
     │   ├── get-tiny-image.ts
     │   ├── gzip-file-as-resource.ts
     │   ├── toggle-logging.ts
     │   ├── toggle-subscriber-updates.ts
     │   ├── trigger-elicitation-request.ts
     │   ├── trigger-long-running-operation.ts
     │   └── trigger-sampling-request.ts
     └── transports
         ├── sse.ts
         ├── stdio.ts
         └── streamableHttp.ts
```

## Key Differences
### Feature Naming
- Naming of tools and prompts are now in `kebab-case` as are files.

### Logging
- Simulated random-leveled logging messages are now opt-in and toggled on or off by a tool.

### Resources 
- The dynamic text/blob resources that formed the resource list are now only accessed via resource templates and are no longer odd/even or upper-bounded. The Dynamic Text Resource and Dynamic Blob Resource templates both take a positive integer and return a dynamically created resource with an included timestamp.
- The listed resources are the files found in the `docs/` folder, and demonstrates exposing actual static filesystem resources.
- Simulated updates of subscribed resources are now opt-in, toggled on or off by a tool.
- A new concept - session resources - is implemented and the `gzip-file-as-resource` tool demonstrates it. A file is compressed by the tool and then added as a session resource, which is available for the duration of the session and is listed along with the other documents when you list resources. Its URI sets it apart from the static resources.

### Prompts
- The prompt `complex_prompt` is now called `args-prompt` because it demonstrates prompts with arguments.
- A new prompt `completable-prompt` demonstrates completable arguments where the selection of the first arg narrows the completions for the second arg.
- The prompt `resource-prompt` now takes a `resourceType` argument (Text or Blob).

### Tools
- Tools are now presented in alphabetical order by name.
- Some tools have been renamed with the goal of making them all verbs.
  - If the tool is meant to get something like an annotated message, its name has a prefix `get-`, e.g., `get-annotated-message`.
  - NOTE: The previous add tool is `get-sum` but `echo` remains as is. These two tools demonstrate the same thing: a tool with an input schema that has a different response based on the input. For this reason, I believe we just should keep one, but they both exist for now.
- The `zip` tool is now `gzip-file-as-resource` and incorporates changes in [this PR](https://github.com/modelcontextprotocol/servers/pull/2831) by @ochafik 
- The structured-content` tools is now `get-structured-content` and provides an enum of city names, returning a different content block for each city.

### Roots
- Although there was code that *should* have done it, the automatic request to a client supporting roots for the initial list was **not** being sent. The listener for `roots/list_changed` was set and so when you added some via the Inspector, it would respond with a `roots/list` request. 
  - The problem was that the call for roots needed to happen after the `notification/initialized` handler was done, otherwise, the request gets lost; the client never gets it.
  - This has been fixed. Screenshot below.
- Now that there should always be a server-cached roots list for each connected client, that is used in the `get-roots-list` tool, unless for whatever reason, the list isn't present when the tool is run, in which case the list is requested.

### Documentation
- `README.md` is minimal and focused on how to launch and use with agents. It contains links to all the other pages in `docs/`.
- `CLAUDE.md` is now `AGENTS.md`
- The files in `docs/` describe Architecture, Project Structure, The Startup Process, Server Features, Extension Points, and How It Works.

### NPM Scripts
* Launcher scripts go through `dist/index.js` rather than pointing to transport manager directly.
  - `start:stdio` 👉 `node dist/index.js stdio`
  - `start:sse` 👉  `node dist/index.js sse`
  - `start:streamableHttp`  👉`node dist/index.js`
* Added `prettier`
  - `prettier:fix`
  - `prettier:check`

## Server Details
<!-- If modifying an existing server, provide details -->
- Server: everything
- Changes to: literally everything 

## Motivation and Context
<!-- Why is this change needed? What problem does it solve? -->
### Use new APIs
The existing implementation of the Everything server uses the low-level `Server` class rather than `McpServer` which does a lot of heavy lifting. As a result, the code is more complicated than it should be. Since this is meant to be an educational resource, a model for how to build servers with the TypeScript SDK, it should use the latest features wherever possible.

### Make it easier to extend
When adding a new feature, such as some of the things we need to add in order to demonstrate the new spec additions, you had to edit multiple places inside the monolithic `everything.ts` file. Add an input schema in one place, a tool description in another, and the tool request handler in a third. Now all of that goes in a single file and an orchestrator registers the tool (or resource, or prompt). Adding other behaviors like simulated logging and resource subscription updates was tough because there was no pattern to follow. Now it is easy to do with those examples in place to follow.

### Add more inline and high-level documentation
Since this server is intended to exercise all of the MCP features, having all the code in a single, monolithic file made it hard to understand and reason about. Further, because of the size of the file, inline documentation was dissuaded as it only made scrolling to a feature area a longer journey. Breaking all the features out into separate files allowed for much better function docs and inline comments. And the system itself is now more fully documented in a set of markdown files, which the server exposes as resources (and the `README.md` links to).

## How Has This Been Tested?
<!-- Have you tested this with an LLM client? Which scenarios were tested? -->
Locally with the Inspector on the branch for support of the new spec's Elicitation Enum shapes https://github.com/modelcontextprotocol/inspector/pull/952. This feature will be in 0.18.0.

## Screenshots 
Most of the tools, prompts, and resources operate in the same fashion as previously. Here are some screenshots showing new behavior.

### Initial `roots/list` request now being sent and list received.
<img width="799" height="374" alt="Screenshot 2025-12-13 at 4 10 19 PM" src="https://github.com/user-attachments/assets/c188c50c-1e4d-4d1e-a45c-cbe7bc9baf37" />

### Resources now include static files and templates for dynamic text and blobs.
<img width="1596" height="631" alt="Screenshot 2025-12-13 at 4 33 22 PM" src="https://github.com/user-attachments/assets/3672d7ba-9d20-4f55-a3e8-9c0c7ad0aba4" />

### New `completable-prompt` demonstrates two stage argument filtering

* Choose department from list
<img width="795" height="491" alt="Screenshot 2025-12-13 at 4 35 19 PM" src="https://github.com/user-attachments/assets/44b53814-7145-434f-84aa-8c25ed05b0f2" />

* Employees shown are only those in selected department
<img width="797" height="498" alt="Screenshot 2025-12-13 at 4 35 28 PM" src="https://github.com/user-attachments/assets/57703e5c-b11d-4f35-8f3c-911368eb9f22" />

* Final result combines both arguments
<img width="802" height="700" alt="Screenshot 2025-12-13 at 4 35 38 PM" src="https://github.com/user-attachments/assets/d4c1e5b1-3471-455e-9fe0-1c85b8a75b39" />

### New `toggle-subscriber-updates` tool

* Simulated Resource Subscription updates are no longer sent automatically, but toggled with a tool that provides more information about what is happening.
<img width="800" height="468" alt="Screenshot 2025-12-13 at 4 47 37 PM" src="https://github.com/user-attachments/assets/c45c2f9b-5a9a-4a85-8c17-7ca7fe6e8cc1" />

### New `toggle-simulated-logging` tool

* Simulated logging is also no longer sent automatically, but instead turned on or off with a tool that provides more information about what is happening.
<img width="801" height="500" alt="Screenshot 2025-12-13 at 4 49 31 PM" src="https://github.com/user-attachments/assets/08a6169b-9f2d-4a69-8df9-48af8b072711" />

### Renamed `gzip-file-as-resource` tool creates session resource
* Resource or resource link is returned when running tool
<img width="798" height="766" alt="Screenshot 2025-12-13 at 4 52 16 PM" src="https://github.com/user-attachments/assets/c0c19e4f-6ae4-4a7d-b0fc-76fe0c7d5c4e" />

* Resource exists in resource list for duration of session with URI that marks it as a session resource rather than static.
<img width="1602" height="678" alt="Screenshot 2025-12-13 at 4 53 15 PM" src="https://github.com/user-attachments/assets/58741fff-05ed-4858-aba2-ec003e0b085d" />

## Breaking Changes
<!-- Will users need to update their MCP client configurations? -->
Yes. Tools and Resources are named differently, but the way that you launch it has not changed, so it shouldn't actually affect anyone.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Documentation update
- [X] Code refactor

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [X] I have read the [MCP Protocol Documentation](https://modelcontextprotocol.io)
- [X] My changes follows MCP security best practices
- [X] I have updated the server's README accordingly
- [ ] I have tested this with an LLM client
- [X] My code follows the repository's style guidelines
- [ ] New and existing tests pass locally
- [X] I have added appropriate error handling
- [ ] I have documented all environment variables and configuration options

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->
This fixes #3080

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs

In src/everything:

* Refactor / move streamableHttp.ts, sse.ts, and stdio.ts to transports/

* Move everything.ts to server/ for reference

* Add server/index.js
  - exports the createServer function
  - import registerTools from tools/index.js
  - in createServer()
    - read instructions.md and include in ServerOptions for McpServer constructor
    - construct McpServer instead of Server
    - call registerTools, passing server

* Add tools/echo.ts
  - define EchoSchema
  - define tool config
  - export addToolEcho function
  - in addToolEcho()
    - register handler for Echo tool

* Add tools/index.ts
  - import addToolEcho
  - export registerTools function
  - in registerTools()
    - call addToolEcho

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Adding dynamic resources

* Add server/index.js
  - import registerResources from resources/index.js
  - in createServer()
    - call registerResources, passing server

* Add resources/dynamic.ts
  - in addDynamicResources()
    - define formatGmtTimestamp to create a time stamp to include in the resource text or encoded blob
    - define parseIndex to ensure the index variable of the URI is a number

* Add resources/index.ts
  - import addDynamicResources
  - export registerResources function
  - in registerResources()
    - call addDynamicResources

* In package.json
  - update the start commands to launch each of the transports properly

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Adding static resources, move server instructions to
the new docs folder, and add code formatting

* Add docs folder

* Add docs/architecture.md which describes the architecture of the project thus far.

* Refactor moved instructions.md to docs/server-instructions.md

* Add resources/static.ts
  - in addStaticResources()
    - read the file entries from the docs folder
    - register each file as a resource (no template), with a readResource function that reads the file and returns it in a contents block with the appropriate mime type and contents
  - getMimeType helper function gets the mime type for a filename
  - readSafe helper function reads the file synchronously as utf-8 or returns an error string

* Add resources/index.ts
  - import addStaticResources
  - export registerResources function
  - in registerResources()
    - call addStaticResources

* In package.json
  - add prettier devDependency
  - add prettier:check script
  - add prettier:fix script
  - in build script, copy docs folder to dist

* All other changes were prettier formatting

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Adding prompts (simple, complex, with completions)

* Add prompts/simple.ts
  - in addSimplePrompt()
    - register a simple prompt with no arguments

* Add prompts/complex.ts
  - in addComplexPrompt()
    - define promptArgsSchema containing a required city arg of type string and an optional state arg of type state
    - register the complex prompt with a prompt callback that combines the city and state into a prompt asking for the weather in that location

* Add prompts/completions.ts
  - in addPromptWithCompletions()
    - define promptArgsSchema containing department and name string fields with completion handlers
    - register the completable prompt with a prompt callback that combines the inputs into a prompt asking to promote the selected name to head of the selected department

* Add prompts/index.ts
  - import addSimplePrompt, addComplexPrompt, and addPromptWithCompletions
  - export registerPrompts function
  - in registerPrompts()
    - call addSimplePrompt
    - call addComplexPrompt
    - call addPromptWithCompletions

* In package.json
  - add prettier devDependency
  - add prettier:check script
  - add prettier:fix script
  - in build script, copy docs folder to dist

* All other changes were prettier formatting

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Updated architecture.md

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Updated architecture.md

* Refactor / renamed all addXPrompt, addXTool, and addXResource functions to registerX...

* Added the add tool

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Updated architecture.md

* Refactor / renamed uris from test://X/resource/Y to demo://resource/X/Y

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Updated architecture.md

* Refactor/renamed resources/dynamic.ts to resources/template.ts
  - refactor/renamed registerDynamicResources to registerResourceTemplates
    - this highlights the more salient fact that we are demonstrating registration of resource templates in this example.
  - exposed the ability to dynamically create the text resources from elsewhere (namely the resource-prompt example

* Added prompts/resource.ts
  - in registerEmbeddedResourcePrompt()
    - register a prompt that takes a resourceId and returns the prompt with the corresponding dynamically created resource embedded

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Ran prettier

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Updated architecture.md

* Refactor/renamed static.ts to file.ts

* Refactor/renamed complex.ts to args.ts

* Refactor/renamed template.ts to templates.ts.

* In resource.ts,
  -  improved registerEmbeddedResourcePrompt to allow selection of blob or text resource type.

* In file.ts
  - refactor/renamed registerStaticResources to registerFileResources to highlight the fact that it is using files as resources.

* In args.ts
  - refactor/renamed registerComplexPrompt to registerArgumentsPrompt to highlight the fact that it is demonstrating prompt arguments.

* Updated inline documentation throughout

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Adding resource subscriptions:

* Updated architecture.md

* In server/index.ts
  - imported Transport, setSubscriptionHandlers,beginSimulatedResourceUpdates, and stopSimulatedResourceUpdates
  - call setSubscriptionHandlers passing server
  - in returned object,
    - refactor/renamed startNotificationIntervals placehodler to clientConnected, which takes a transport argument and calls beginSimulatedResourceUpdates, passing the transport
    - replaced cleanup placeholder with a function that takes an optional sessionId and calls stopSimulatedResourceUpdates, passing the sessionId

* In sse.ts, stdio.ts, and streamableHttp.ts
  - when transport is connected, called clientConnect, passing transport
  - when disconnecting, called cleanup, passing sessionId

* Added subscriptions.ts
  - tracks subscriber session id lists by URI
  - tracks transport by session id
  - tracks subscription update intervals by sessionId
  - in setSubscriptionHandlers
    - set request handlers for SubscribeRequestSchema and UnsubscribeRequestSchema
  - in beginSimulatedResourceUpdates
    - starts an interval to send updates to the transport for all subscribed resources
  - in stopSimulatedResourceUpdates
    - removes intervals and transport for gien session id

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* In subscriptions.ts
  - updated inline doc

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Adding simulated logging and refactoring subscriptions to not need to track transports

* Updated architecture.md

* In server/index.ts
  - remove import of Transport
  - import beginSimulatedLogging and stopSimulatedLogging
  - in clientConnected()
    - change argument to sessionId? instead of transport
    - add call to beginSimulatedLogging
    - send server and sessionId to beginSimulatedResourceUpdates and beginSimulatedLogging
  - in cleanup()
    - add call to stopSimulatedLogging passing sessionId

* Added server/logging.ts
  - Initialize logsUpdateIntervals to Map session ID to the interval for sending logging messages to the client
  - in beginSimulatedLogging()
    - create an array of logging meesages, customized with the sessionId if present
    - if the interval for the sessionId hasn't been set, create one, calling server.sendLoggingMessage with a random message to the client each time the interval elapses
  - in stopSimulatedLogging()
    - if a logging interval exists for the sessionId, clear it and remove it

* In subscriptions.ts
  - remove import of Transport
  - remove transports map
  - in beginSimulatedResourceUpdates()
    - change arguments to server and sessionId
    - check for the subsUpdateInterval for the session
    - remove all transport storage and interaction
    - instead use the server to send the notification
 - in stopSimulatedResourceUpdates()
   - remove management of transports map

* In sse.ts and streamableHttp.ts
  - when calling clientConnected, pass sessionId instead of transport

* In stdio.ts,
- when calling clientConnected, pass nothing instead of transport

* In subscriptions.ts
  - updated inline doc

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Added tools to toggle simulated logging and resource updates on and off rather than have them start immediately upon connection

* Updated architecture.md

* In server/index.ts
  - remove import of beginSimulatedResourceUpdates and beginSimulatedLogging
  - remove clientConnected from createServer factory result

* In tools/index.ts
  - import registerToggleLoggingTool and registerToggleSubscriberUpdatesTool
  - in registerTools
    - call registerToggleLoggingTool and registerToggleSubscriberUpdatesTool

* In logging.ts
  - in beginSimulatedLogging
    - refactor extract inline interval callback into function sendSimulatedLoggingMessage
  - call sendSimulatedLoggingMessage right away to send the first message
  - supply sendSimulatedLoggingMessage as interval callback

* In subscriptions.ts
  - remove import of Transport
  - remove transports map
  - in beginSimulatedResourceUpdates()
    - change arguments to server and sessionId
    - check for the subsUpdateInterval for the session
    - remove all transport storage and interaction
    - instead use the server to send the notification
 - in stopSimulatedResourceUpdates()
   - remove management of transports map

* In stdio.ts, sse.ts, and streamableHttp.ts
  - remove destructure and calling of clientConnected

* Added tools/toggle-logging.ts
  - registers a tool that
    - takes no arguments
    - tracks clients that have been enabled by session id in a set
    - if client isn't enabled,
       - calls beginSimulatedLogging
       - adds session id to client set
    - else
      - calls stopSimulatedLogging
      - deletes session id from client set
    - returns a message explaining what was done including what to expect when logging is enabled

 * Added tools/toggle-subscriber-updates.ts
   - registers a tool that
     - takes no arguments
     - tracks clients that have been enabled by session id in a set
     - if client isn't enabled,
        - calls beginSimulatedResourceUpdates
        - adds session id to client set
     - else
       - calls stopSimulatedResourceUpdates
       - deletes session id from client set
     - returns a message explaining what was done including what to expect when logging is enabled

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Added long-running-operation tool and improved comments in all tools

* Updated architecture.md

* In tools/
  - add.ts
  - echo.ts
  - toggle-logging.ts
  - toggle-subscriber-updates.ts
    - Add better function and inline docs

* Added tools/long-running-operation.ts
  - similar implementation as in everything v1

* In tools/index.ts
  - import registerLongRunningOperationTool
  - in registerTools
    - registerLongRunningOperationTool

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Added print-env, and sampling-request tools

* Updated architecture.md

* In tools/index.ts
  - import registerPrintEnvTool and registerSamplingRequestTool
  - in registerTools
    - call registerPrintEnvTool and registerSamplingRequestTool

* Added tools/print-env.ts
  - registers a tool that takes no args and returns the environment variables

* Added tools/sampling-request
  - registers a tool that
    - takes prompt and maxTokens args
    - sends client a sampling request
    - returns the client response in the result

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Updated architecture.md

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* In sampling-request.ts
  - renamed SampleLLMSchema to SamplingRequestSchema

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Added git-tiny-image.ts
  - registers a tool that returns a tiny MCP logo

* In all other tools, updated function docs

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Adding the annotated message tool

* Updated architecture.md

* Added annotated-message.ts
  - registers a tool that returns a message with annotations on different content types

* In package.json
  - updated TS SDK version

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Adding the annotated message tool

* Updated architecture.md

* In annotated-message.ts
  - prettier

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Adding the get-resource-reference tool

* Updated architecture.md

* In prompts/resource.ts
  - Refactor/extracted the prompt argument completers into exported functions in resources/templates.ts
  - Refactor/extracted BLOB_TYPE, TEXT_TYPE, and resourceTypes into exported constants in resources/templates.ts as RESOURCE_TYPE_BLOB, RESOURCE_TYPE_TEXT, and RESOURCE_TYPES
  - In resources/templates.ts
    - refactor renamed index to resourceId throughout for consistency with prompts and tool references
* Added tools/get-resource-reference.ts
  - Registers the 'get-resource-reference' tool with the provided McpServer instance.
  - uses enum and number schema for tools to provide resourceType and resourceId arguments. Completables don't work for tool arguments.
  - Returns the corresponding dynamic resource
* In tools/index.ts
  - imported registerGetResourceReferenceTool
  - in registerTools
    - called registerGetResourceReferenceTool

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Adding the get-resource-reference and get-resource-reference tools

* Updated architecture.md

* Added get-resource-links.ts
  - Registers the 'get-resource-reference' tool with the provided McpServer instance.
  - The registered tool retrieves a specified number of resource links and their metadata.

* In get-resource-reference.ts
  - fixed tool description

* In tools/index.ts
  - import registerGetResourceLinksTool
  - in registerTool
    - call registerGetResourceLinksTool

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Refactor/rename `print-env` tool to `get-env`

* Updated architecture.md

* Refactor rename print-env.ts to get-env.ts

* In tools/index.ts
  - reorder tools alphabetically

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* In get-env.ts
  - update tool description

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Updated architecture.md

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Adding get-structured-content tool

* Updated architecture.md

* added get-structured-content.ts
  - Registers the 'get-structured-content' tool with the provided McpServer instance.
  - The registered tool processes incoming arguments using a predefined input schema,
    generates structured content with weather information including temperature,
    conditions, and humidity, and returns both backward-compatible content blocks
    and structured content in the response.

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

For tools where we seek to get some response from the server, renamed as an action, e.g., "Get Sum" rather than "Add" or "Get Annotated Message" rather than "Annotated Message", so that it's clear what the intent of the tool is in a quick review.

* Updated architecture.md

* Refactor/renamed add.ts to get-sum.ts
* Refactor/renamed annotated-message.ts to get-annotated-message.ts
* In tools/index.ts
  - sorted presentation order

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Updated architecture.md

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Updated architecture.md

* Refactor/renamed sampling-request.ts to get-sampling-request.ts

* In tools/index.ts
  - sorted presenation order

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

[WIP] Adding Gzip File as Resource tool

* Updated architecture.md

* In resources/files.ts
  - remove 'Docs: ' from resource name

* In resources/templates.ts
  - use blob instead of text in resource content

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

[WIP] Adding Gzip File as Resource tool

* Updated architecture.md

* Added gzip-file-as-resource.ts
  - imports getSessionResourceURI and  registerSessionResource from session.ts
  - exports registerGZipFileAsResourceTool
    - the registered tool
      - validates the input URI
      - fetches the file safely
      - compresses it
      - creates and registers the resource
      - returns resource or resource link

* In tools/index.ts
  - import registerGZipFileAsResourceTool
  - in registerTools,
    - call registerGZipFileAsResourceTool passing server

* Added resources/session.ts
  - getSessionResourceURI gets a uri to the specified name
  - registerSessionResource registers the session-scoped resource and returns a resource link

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Updated architecture.md

* Refactor/renamed get-sampling-request.ts to trigger-sampling-request.ts
  - use trigger instead of get throughout

* In tools/index.ts
  - sorted display order

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Adding Trigger Elicitation Request tool

* Updated architecture.md

* Added trigger-elicitation-request.ts
  - registerTriggerElicitationRequestTool
    - registered tool sends an elicitation request that exercises all supported field types

* In tools/index.ts
  - imports registerTriggerElicitationRequestTool
  - in registerTools
    - call registerTriggerElicitationRequestTool passing server

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Adding Trigger Elicitation Request and Get Roots List tools

* Updated architecture.md

* Added roots.ts
  - tracks roots by sessionId
  - setRootsListChangedHandler
    - listens for roots changed notification from the client
      - updates the roots map by sessionId
      - sends log notification or error to the client

* In server/index.ts
  - import setRootsListChangedHandler
  - in clientConnected callback
    - call setRootsListChangedHandler passing server and sessionId

* In sse.ts, stdio.ts, and streamableHttp.ts
  - receive clientConnected from server factory
  - call clientConnected when server is connected to transport
* Added get-roots-list.ts
  - registerGetRootsListTool
    - Registers the 'get-roots-list' tool with the given MCP server.

* Added trigger-elicitation-request.ts
  - registerTriggerElicitationRequestTool
    - registered tool sends an elicitation request that exercises all supported field types

* In tools/index.ts
  - imports registerTriggerElicitationRequestTool and registerGetRootsListTool
  - in registerTools
    - call registerTriggerElicitationRequestTool and registerGetRootsListTool, passing server

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Adding Trigger Elicitation Request and Get Roots List tools

* Updated architecture.md

* Added roots.ts
  - tracks roots by sessionId
  - setRootsListChangedHandler
    - listens for roots changed notification from the client
      - updates the roots map by sessionId
      - sends log notification or error to the client

* In server/index.ts
  - import setRootsListChangedHandler
  - in clientConnected callback
    - call setRootsListChangedHandler passing server and sessionId

* In sse.ts, stdio.ts, and streamableHttp.ts
  - receive clientConnected from server factory
  - call clientConnected when server is connected to transport
* Added get-roots-list.ts
  - registerGetRootsListTool
    - Registers the 'get-roots-list' tool with the given MCP server.

* Added trigger-elicitation-request.ts
  - registerTriggerElicitationRequestTool
    - registered tool sends an elicitation request that exercises all supported field types

* In tools/index.ts
  - imports registerTriggerElicitationRequestTool and registerGetRootsListTool
  - in registerTools
    - call registerTriggerElicitationRequestTool and registerGetRootsListTool, passing server

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Finalized Roots list changed handling and initial request. Final fit and finish work.

* Updated architecture.md
  - Added links to other docs
  - Refactor/extracted sections into extension.md, features.md, how-it-works.md, startup.md, and structure.md

* Removed everything.ts
  - all features are ported

* In roots.ts
  - refactor/renaned setRootsListChangedHandler to syncRoots
  - refactor handler logic to requestRoots function
  - Calls for roots list directly to get initial list

* In server/index.ts
  - import setRootsListChangedHandler
  - in clientConnected callback
    - call setRootsListChangedHandler passing server and sessionId

* In sse.ts, stdio.ts, and streamableHttp.ts
  - update inline and function docs

* In index.ts,
  - updated usage output

* In server/index.ts
  - refactor/extracted readInstructions to resources/index.ts
  - defined ServerFactoryResponse response type

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Updated README.md

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Updated docs

* Refactor/renamed CLAUDE.md to AGENTS.md
* Updated instructions.md and startup.md

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

Updated structure.md

cliffhall committed:
[WIP] Refactor everything server to be more modular and use recommended APIs.

* Updated docs

* Refactor/renamed toggle-logging.ts to toggle-simulated-logging.ts
  - refactor/renamed registerToggleLoggingTool to registerToggleSimulatedLoggingTool

cliffhall committed:
Merge branch 'main' into new-everything-server

# Conflicts:
#	Removed src/everything/everything.ts

cliffhall committed:
Updated structure.md

cliffhall committed:
Refactor-renamed long-running-operation.ts to trigger-long-running-operation.ts

cliffhall committed:
Updated AGENTS.md with guidance on tool naming

cliffhall committed:
Updated package.json and ran prettier

cliffhall committed:
Update README.md and doc in server/index.ts

cliffhall committed:
Update doc in server/index.ts

cliffhall committed:
Update doc in server/index.ts

cliffhall committed:
Update doc in tools/toggle-subscriber-updates.ts

cliffhall committed:
Update doc in tools/trigger-elicitation-request.ts

cliffhall committed:
Update doc in tools/trigger-long-running-operation.ts

cliffhall committed:
Update doc in tools/trigger-elicitation-request.ts

cliffhall committed:
Update doc in tools/trigger-elicitation-request.ts and trigger-long-running-operation.ts

cliffhall committed:
Update doc in tools/toggle-subscriber-updates.ts

cliffhall committed:
Update doc in tools/toggle-subscriber-updates.ts

cliffhall committed:
Update doc in tools/toggle-subscriber-updates.ts

cliffhall committed:
Update doc in tools/gzip-file-as-resource.ts

cliffhall committed:
Use roots cache in get-roots-list tool
* In get-roots-list.ts, query the server's cache of the latest roots from the client and only request the list from the client if it doesn't exist.
* In roots.ts, export the roots map

cliffhall committed:
Use roots cache in get-roots-list tool
* In get-roots-list.ts, query the server's cache of the latest roots from the client and only request the list from the client if it doesn't exist.
* In roots.ts, export the roots map

cliffhall committed:
Run pretter:fix

cliffhall committed:
Updated doc in get-resource-reference.ts

cliffhall committed:
Updated doc in get-env.ts

cliffhall committed:
Updated doc in get-annotated-message.ts

cliffhall committed:
Updated doc in echo.ts

cliffhall committed:
Updated doc in trigger-sampling-request.ts

cliffhall committed:
Updated doc in all tools

cliffhall committed:
Updated doc in all tools

cliffhall committed:
Cache roots in get-roots-list.ts if list was not already present and had to be requested.

olaservo commented on src/everything/server/index.ts:
```diff
@@ -0,0 +1,83 @@
+      version: "2.0.0",
+    },
+    {
+      capabilities: {
```
Should this declare `listChanged: true` for tools/resources/prompts?

olaservo commented on src/everything/server/index.ts:
```diff
@@ -0,0 +1,83 @@
+        prompts: {},
+        resources: {
+          subscribe: true,
+        },
```
Should we add the `completions: {}` capability since the server supports completions?

cliffhall replied:
TL;dr The completions capability doesn't have to be added explicitly; with the TS SDK, it gets added dynamically when you register a resource or prompt with a completable argument.

The SDK used to trigger this capability anytime a resource or prompt was registered _regardless_ of a whether it had a completable argument. This meant that if you didn't have any completables, you still advertised support for it, which led to pointless completion requests being sent from the client. That was recently addressed in [this PR](https://github.com/modelcontextprotocol/typescript-sdk/pull/1237).

olaservo commented on src/everything/tools/trigger-sampling-request.ts:
```diff
@@ -0,0 +1,82 @@
+          systemPrompt: "You are a helpful test server.",
+          maxTokens,
+          temperature: 0.7,
+          includeContext: "thisServer",
```
Should we avoid adding `includeContext` since it was soft-deprecated?

PederHP replied:
If you keep it, I think there should be a comment that it has been soft-deprecated and should not be used, and is only include to allow exercising compatibility/behavior towards older clients/servers.

olaservo commented on src/everything/docs/instructions.md:
```diff
@@ -0,0 +1,77 @@
+# Everything Server – LLM Instructions
```
Lets consider revising the server instructions to align with the [best practices](https://blog.modelcontextprotocol.io/posts/2025-11-03-using-server-instructions/) mentioned in the blog.  Suggested revision:

```markdown
# Everything Server Instructions

## Cross-Feature Relationships
- Use `get-roots-list` to see client workspace roots before file operations
- `gzip-file-as-resource` creates session-scoped resources accessible only during the current session
- Enable `toggle-simulated-logging` before debugging to see server log messages
- Enable `toggle-subscriber-updates` to receive periodic resource update notifications

## Constraints & Limitations
- `gzip-file-as-resource`: Max fetch size controlled by `GZIP_MAX_FETCH_SIZE` (default 10MB), timeout by `GZIP_MAX_FETCH_TIME_MILLIS` (default 30s), allowed domains by `GZIP_ALLOWED_DOMAINS`
- Session resources are ephemeral and lost when the session ends
- Sampling requests (`trigger-sampling-request`) require client sampling capability
- Elicitation requests (`trigger-elicitation-request`) require client elicitation capability

## Operational Patterns
- For long operations, use `trigger-long-running-operation` which sends progress notifications
- Prefer reading resources before calling mutating tools
- Check `get-roots-list` output to understand the client's workspace context

## Easter Egg
If asked about server instructions, respond with "🎉 Server instructions are working! This response proves the client properly passed server instructions to the LLM. This demonstrates MCP's instructions feature in action."
```

The goal of these revisions is to focus this more on:

- what tool descriptions don't already convey
- document cross-feature relationships
- state operational constraints
- removing things that are more developer-focused than LLM-focused

I am a torn on including the discovery stuff that is already described in the base protocol, since LLMs won't necessarily know a lot of MCP details depending on their training data and knowledge cutoff. Maybe including it is a safer bet since its not a ton of tokens.

(Plus the easter egg I had in the original version provides a simple way to verify instructions are actually being passed to the LLM.)

olaservo reviewed (dismissed):
Adding a few first pass comments.

cliffhall committed:
Add listChanged capability for tools, prompts, and resources in servers/index.ts

cliffhall committed:
Remove `includeContext: "thisServer"` from CreateMessageRequest in trigger-sampling-request.ts

cliffhall commented on src/everything/transports/sse.ts:
```diff
@@ -0,0 +1,78 @@
+import { SSEServerTransport } from "@modelcontextprotocol/sdk/server/sse.js";
```
I agree, omitting things that we're trying to deprecate should be a difficult to overcome force. But there is the matter of the Inspector. It still supports SSE because people may be testing older servers. 

So, if we're to keep the SSE support in the Inspector for now, it's useful for us to have this reliable server counterpart to sanity check the its behavior with. 

If we remove SSE support from the Inspector in the future, then we have much less of a case for keeping it here either.

cliffhall committed:
Updated server instructions.md. See https://github.com/modelcontextprotocol/servers/pull/3121#discussion_r2616651611

cliffhall committed:
Updated server instructions.md. See https://github.com/modelcontextprotocol/servers/pull/3121#discussion_r2616651611

cliffhall committed:
Updated server instructions.md. See https://github.com/modelcontextprotocol/servers/pull/3121#discussion_r2616651611

cliffhall committed:
Demonstrate registration of tools conditioned upon client capability support. Also, obviated need for clientConnected callback to pass sessionId because we defer initial fetching of roots happens when you run the get-roots-list tool.

* In how-it-works.md,
  - added a section on conditional tool registration

* In server/index.ts
  - import registerConditionalTools
  - in an oninitialized handler for the server, call registerConditionalTools
  - removed clientConnected from ServerFactoryResponse and all mentions in docs

* In tools/index.ts
  - export a registerConditionalTools function
  - refactor/move calls to registerGetRootsListTool, registerTriggerElicitationRequestTool, and registerTriggerSamplingRequestTool out of registerTools and into registerConditionalTools

* In server/roots.ts
  - only act if client supports roots
  - remove setInterval from call to requestRoots. It isn't happening during the initialze handshake anymore, so it doesn't interfere with that process if called immediaately

* In get-roots-list.ts, trigger-elicitation-request.ts, and trigger-sampling-request.ts,
  - only register tool if client supports capability

* Throughout the rest of the files, removing all references to `clientConnected`

cliffhall committed:
Get roots unconditionally when syncRoots is called

* In server/roots.ts
  - in syncRoots, if roots are supported fetch them

* In get-roots-list.ts,
  - Don't import and inspect current roots map, just call syncRoots to get the list.

cliffhall committed:
Merge branch 'main' into new-everything-server

cliffhall committed:
Sync roots at startup. This was deferred to list_roots tool is used, but I'm putting it back.

The syncRoots call should be idempotent, requesting roots if they haven't been yet for the session, but always retuning the cached roots otherwise. That could be deferred but setting the handler for roots_list changed note should not.

* In server/roots.ts
  - only set the notification handler and call for initial roots list if the roots aren't already cached for this client.

* In server/index.ts
  - in the oninitialized handler
    - get the sessionId from the transport
    - set a 350ms timeout to call syncRoots with the server and sessionId
      - this delay cause it to run after the `notifications/initialized` handler finishes, otherwise, the request gets lost.

* All other changes attributable to prettier

cliffhall committed:
Sync roots at startup. This was deferred to list_roots tool is used, but I'm putting it back.

* In server/roots.ts
  - Updated function doc

cliffhall committed:
Merge branch 'main' into new-everything-server

domdomegg commented on src/everything/prompts/args.ts:
```diff
@@ -0,0 +1,41 @@
+
+  // Register the prompt
+  server.registerPrompt(
+    "args-prompt",
```
nit: The convention I've generally seen for prompts and tool names is `snake_case` - is there a reason we're using kebab case? Seems good to follow conventions where we can in case agents use this to see how to implement particular features etc. and pick up on patterns.

cliffhall replied:
The `snake_case` convention mostly comes from Python server development, because the convention matches Python idioms. The point of deviating here is that it illustrates that snake is a convention, but tool names can be whatever you want them to be within the allowed parameters. A `kebab-case` primitive can line up with the file name that registers making for a single naming strategy. Same benefit of having one case pattern throughout the project, but with TS friendly idioms. It's a design choice one can make.

domdomegg commented on src/everything/resources/templates.ts:
```diff
@@ -0,0 +1,211 @@
+const parseResourceId = (uri: URL, variables: Record<string, unknown>) => {
+  const uriError = `Unknown resource: ${uri.toString()}`;
+  if (
+    uri.toString().startsWith(textUriBase) &&
```
```suggestion
    uri.toString().startsWith(textUriBase) ||
```

I think this is the intention?

domdomegg reviewed (dismissed):
Generally LGTM, very excited about this! Thank you :)

cliffhall committed:
Update src/everything/AGENTS.md

Co-authored-by: adam jones <domdomegg+git@gmail.com>

olaservo reviewed (dismissed):
This looks good to me - I think we can merge this and start adding the other couple of examples we've mentioned.

I also did one more claude-driven Playwright test run for all the features for each transport using inspector v 0.18.0, along with manual testing for stdio and didn't run into any errors.

cliffhall committed:
Remove console log statements
